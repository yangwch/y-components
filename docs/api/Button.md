---
category: api
type: 数据展示
title: Button
toc: content
---
## Button

按钮的属性说明如下：

| 属性     | 说明                                                                                                                                  | 类型                                                                          | 默认值      |
| -------- | ------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------- | ----------- |
| danger   | 设置危险按钮                                                                                                                          | boolean                                                                       | false       |
| disabled | 按钮失效状态                                                                                                                          | boolean                                                                       | false       |
| ghost    | 幽灵属性，使按钮背景透明                                                                                                              | boolean                                                                       | false       |
| htmlType | 设置 `button` 原生的 `type` 值，可选值请参考 [HTML 标准](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#attr-type) | string                                                                        | `button`  |
| size     | 设置按钮大小                                                                                                                          | `large` \| `middle` \| `small`                                          | `middle`  |
| type     | 设置按钮类型                                                                                                                          | `primary` \| `ghost` \| `dashed` \| `link` \| `text` \| `default` | `default` |
| onClick  | 点击按钮时的回调                                                                                                                      | (event) => void                                                               | -           |

支持原生 button 的其他所有属性。

## Button.Group

|   属性   | 说明           | 类型                                 | 默认值     |
| :-------: | -------------- | ------------------------------------ | ---------- |
|   size   | 设置按钮大小   | `large` \| `middle` \| `small` | `middle` |
|   style   | 设置 Group样式 | CSSProperties                        | -          |
| className | 设置class样式  | string                               | -          |
