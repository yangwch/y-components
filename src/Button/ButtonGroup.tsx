import classNames from 'classnames';
import React, { CSSProperties } from 'react';
import { settings } from '../_utils/global';
import { sizeClassNameMap, SizeType } from '../_utils/size';
import SizeContext from './SizeContext';

interface ButtonGroupProps {
  /**
   * @description 按钮尺寸："small" | "middle" | "large"
   */
  size?: SizeType;
  children?: React.ReactNode;
  /**
   * @description 样式
   * @type React.CSSProperties
   */
  style?: CSSProperties;
  /**
   * @description 类名
   */
  className?: string;
}
const groupCls = `${settings.prefix}-btn-group`;

function ButtonGroup(props: ButtonGroupProps) {
  const { size: customizeSize, children, className, style } = props;
  const classes = classNames(groupCls, className, {
    [`${groupCls}-${customizeSize}`]: `${customizeSize}` in sizeClassNameMap,
  });
  return (
    <SizeContext.Provider value={customizeSize}>
      <div className={classes} style={style}>
        {children}
      </div>
    </SizeContext.Provider>
  );
}

export default ButtonGroup;
