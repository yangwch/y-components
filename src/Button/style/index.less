@import '../../style/themes/index.less';
@import './mixin.less';

@btn-prefix-cls: ~'@{prefix}-btn';
@btn-group-cls: ~'@{prefix}-btn-group';

.@{btn-prefix-cls} {
  .btn();
  .btn-default();

  >span {
    display: inline-block;
  }

  &-primary {
    .btn-primary();
  }

  &-danger {
    .btn-danger();
  }

  &-dashed {
    .btn-dashed();
  }

  &-ghost {
    .btn-ghost();
  }
}

.@{btn-group-cls} {
  >.@{btn-prefix-cls} {
    border-radius: 0;
    border-left-width: 0;

    &:first-child {
      border-left-width: @btn-border-width;
      border-radius: @btn-border-radius-base 0 0 @btn-border-radius-base;
    }
    

    &:last-child {
      border-right-width: @btn-border-width;
      border-radius: 0 @btn-border-radius-base @btn-border-radius-base 0;
    }

    &:hover,
    &:focus,
    &:active {
      &:not(:first-child)::before {
        position: absolute;
        display: block;
        position: absolute;
        z-index: 1;
        top: -1px;
        left: -1px;
        display: block;
        box-sizing: content-box;
        width: @btn-border-width;
        height: 100%;
        padding: 1px 0;
        background-color: @primary-color-hover;
        transition: background-color .3s;
        content: "";
      }
    }
  }
}