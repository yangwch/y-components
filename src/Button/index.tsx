import classNames from 'classnames';
import React from 'react';
import { settings } from '../_utils/global';
import { sizeClassNameMap, SizeType } from '../_utils/size';
import { tuple } from '../_utils/type';
import ButtonGroup from './ButtonGroup';
import './style/index.less';

const cls_prefix = settings.prefix;

const ButtonTypes = tuple('default', 'primary', 'ghost', 'dashed', 'link', 'text');

export type ButtonType = (typeof ButtonTypes)[number];

interface BaseButtonProps {
  /**
   * @description 按钮类型："default" | "primary" | "ghost" | "dashed" | "link" | "text"
   */
  type?: ButtonType;
  children?: React.ReactNode;
  /**
   * @description 设置危险按钮
   */
  danger?: Boolean;
  /**
   * @description 是否禁用
   * @default 'false'
   */
  disabled?: Boolean;
  /**
   * @description 尺寸："small" | "middle" | "large"
   * @default 'middle'
   */
  size?: SizeType;
  /**
   * @description 类名
   */
  className?: string;
}
const ButtonHTMLTypes = tuple('submit', 'button', 'reset');
export type ButtonHTMLType = (typeof ButtonHTMLTypes)[number];

export type NativeButtonProps = {
  htmlType?: ButtonHTMLType;
  /**
   * @description 点击事件
   */
  onClick?: React.MouseEventHandler<HTMLElement>;
} & BaseButtonProps &
  Omit<React.ButtonHTMLAttributes<any>, 'type' | 'onClick'>;

export type ButtonProps = Partial<NativeButtonProps>;

interface CompoundedComponent
  extends React.ForwardRefExoticComponent<ButtonProps & React.RefAttributes<HTMLElement>> {
  Group: typeof ButtonGroup;
}

const Button = React.forwardRef(function Button(
  props: ButtonProps,
  ref: React.ForwardedRef<HTMLButtonElement>,
) {
  const {
    type = 'default',
    htmlType,
    children,
    danger = false,
    disabled,
    size,
    className,
    ...rest
  } = props;

  const clsNames = classNames(
    {
      [`${cls_prefix}-btn`]: true,
      [`${cls_prefix}-btn-${type}`]: type !== 'default',
      [`${cls_prefix}-btn-danger`]: !!danger,
      [`${cls_prefix}-btn-dashed`]: type === 'dashed',
      [`${cls_prefix}-btn-ghost`]: type === 'ghost',
      [`${cls_prefix}-btn-${sizeClassNameMap[size || 'middle'] || ''}`]: size !== undefined,
    },
    className,
  );
  return (
    <button type={htmlType} className={clsNames} disabled={!!disabled} ref={ref} {...rest}>
      {children}
    </button>
  );
}) as CompoundedComponent;

Button.Group = ButtonGroup;

export { Button, ButtonGroup };
