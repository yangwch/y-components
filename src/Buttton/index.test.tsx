import React from 'react';
import { mountTest } from '../../test/shared/mountTest';

import Button from './index';

describe('Button', () => {
  mountTest(() => <Button />);
  mountTest(() => <Button>Default</Button>);
  mountTest(() => <Button type="primary">Primary</Button>);
  mountTest(() => <Button disabled>Disabled</Button>);
  mountTest(() => <Button type="dashed">Dashed</Button>);
  mountTest(() => <Button danger>Danger</Button>);
  mountTest(() => <Button type="ghost">Ghost</Button>);

  mountTest(() => <Button size="large">Default</Button>);
  mountTest(() => <Button size="large" type="primary">Primary</Button>);
  mountTest(() => <Button size="large" disabled>Disabled</Button>);
  mountTest(() => <Button size="large" type="dashed">Dashed</Button>);
  mountTest(() => <Button size="large" danger>Danger</Button>);
  mountTest(() => <Button size="large" type="ghost">Ghost</Button>);
  
});
