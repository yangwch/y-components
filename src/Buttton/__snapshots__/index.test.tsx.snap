// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button render correctly 1`] = `
<button
  class="yy-btn"
>
  Test
</button>
`;
